### PR TITLE
feat: add `head`, `sample` and axis 0 slicing

### PR DESCRIPTION
**Pull request by martindurant**
prelude to slicing on axis 0

### CODE REVIEW

**Comment by martindurant** [comment]
@douglasdavis , I reckon this is a pretty easy one to allow in quickly. I don't know how effective it would be at culling massive graphs, but extra work on that doesn't belong here.

**Comment by douglasdavis** [comment]
yep looks good to me!

**Comment by martindurant** [comment]
@douglasdavis , added an attempt at mocking a materialised layer for axis 0 slicing, which makes column optimisations pass. It's probably brittle! We should come up with a set of tests first, then fix problems. This will also apply to repartition, though.

Also, we need to decide whether head() and sample() should be dak functions only, as @jpivarski suggested, or if we keep one or both as they are. I think having them as functions is fine, but there is an argument for head() (only) as a method also.

**Comment by douglasdavis** [comment]
> @douglasdavis , added an attempt at mocking a materialised layer for axis 0 slicing, which makes column optimisations pass. It's probably brittle! We should come up with a set of tests first, then fix problems. This will also apply to repartition, though.

great! I'll take a look and test it out

> Also, we need to decide whether head() and sample() should be dak functions only, as @jpivarski suggested, or if we keep one or both as they are. I think having them as functions is fine, but there is an argument for head() (only) as a method also.

I think we decided to make `head` a method (the precedent set by dataframe is huge here). I would argue that `sample` should probably be treated in the same way (`pandas.DataFrame.sample` being a method); though NumPy does have a function for random sampling (`numpy.random.choice`)

**Comment by martindurant** [comment]
`assert 0.3 < (len(arr) / len(daa)) < 0.7` - I should calculate what the actual chance of this failing is, since it's using random probabilities. Perhaps even better, I should change sample() to allow to set the numpy random seed? I'm not sure that will work when the tasks can be running simultaneously.

File: src/dask_awkward/lib/core.py [hunk 1]
```diff
@@ -872,6 +872,69 @@ def _getitem_outer_int(self, where: int | tuple[Any, ...]) -> Any:
...
+                rest,
+            )
+            outpart += 1
+            remainder += (self.divisions[i + 1] - self.divisions[i]) % step
+            divisions.append(
+                (self.divisions[i + 1] - self.divisions[i]) // step + divisions[-1]
+            )
+            remainder = remainder % step
+        hlg = HighLevelGraph.from_collections(name, dask, dependencies=[self])
+        hlg.layers[name].mock = mock_materialized_layer(hlg.layers[name], self.name)
```
Comment (douglasdavis) [inline src/dask_awkward/lib/core.py:929]: this attach-a-free-function-as-instance-method is clever! I was thinking we would need to subclass `MaterialzedLayer` and add a `mock` method

Comment (martindurant) [reply src/dask_awkward/lib/core.py:929]: I was feeling burned from making layers, but yes that would work of course. It's not actually a method, just an attribute here (making a true dynamic method is possible too).

Comment (martindurant) [reply src/dask_awkward/lib/core.py]: done

**Comment by martindurant** [comment]
(when this is in, repartition should use the same mock)

File: src/dask_awkward/lib/inspect.py [hunk 2]
```diff
@@ -78,3 +79,27 @@ def necessary_columns(*args: Any, traverse: bool = True) -> dict[str, list[str]]
...
+
+
+def sample(arr, factor: int | None = None, probability: float | None = None):
+    """Decimate the data to a smaller number of rows
+
+    Must give either factor or probability
+
+    Parameters
+    ----------
+    arr: dask-awkward array
```
Comment (douglasdavis) [inline src/dask_awkward/lib/inspect.py:91]:
```suggestion
    arr : dask_awkward.Array
        Array collection to sample
```

File: src/dask_awkward/lib/inspect.py [hunk 3]
```diff
@@ -78,3 +79,27 @@ def necessary_columns(*args: Any, traverse: bool = True) -> dict[str, list[str]]
...
+    factor: if given, every Nth row will be kept. The counting restarts for each
+        partition, so reducing the row count by an exact factor is not guaranteed
```
Comment (douglasdavis) [inline src/dask_awkward/lib/inspect.py:93]:
```suggestion
    factor : int, optional
        if given, every Nth row will be kept. The counting restarts for each
        partition, so reducing the row count by an exact factor is not guaranteed
```

File: src/dask_awkward/lib/inspect.py [hunk 4]
```diff
@@ -78,3 +79,27 @@ def necessary_columns(*args: Any, traverse: bool = True) -> dict[str, list[str]]
...
+    probability: a number between 0 and 1, giving the chance of any particular
+        row surviving. For instance, for probability=0.1, roughly 1-in-10
+        rows will remain.
```
Comment (douglasdavis) [inline src/dask_awkward/lib/inspect.py:96]:
```suggestion
    probability : float, optional
        a number between 0 and 1, giving the chance of any particular
        row surviving. For instance, for probability=0.1, roughly 1-in-10
        rows will remain.
```

**Review by douglasdavis — COMMENTED** [review]
a couple of docstring nitpicks but other than that this looks good to me

**Comment by douglasdavis** [comment]
I know there exists a sphinx plugin to handle this- I'd like to eventually get all sphinx types from the typehints, it's just not high on my todo list right now :)

**Comment by douglasdavis** [comment]
Do you have anything else in mind for this PR or good to go?

**Comment by martindurant** [comment]
> Do you have anything else in mind for this PR or good to go?

I feel like the mock will need revisiting, but maybe we can make more extensive test cases for it out of real workflows after merging. I don't have a clear idea of how it can be improved now, since I don't know how it will break.

**Pull request merged**